### PR TITLE
Development of the flatten plugin

### PR DESCRIPTION
**Pull request by pbailly**
[[ch62192]](https://app.clubhouse.io/dataiku/story/62192/flatten-table-plugin)
Add a plugin in the code that allows users to flatten hierarchical data retrieved from Big Query.

Associated PR: https://github.com/dataiku/dip/pull/12461

### CODE REVIEW

File: tests/python/integration/pytest.ini [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[pytest]
```
Comment (pbailly) [inline tests/python/integration/pytest.ini]: Tests are legacy code from the original plugin. These need to change.

File: custom-recipes/big-query-unnest/recipe.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
+    "meta": {
+        "label": "BigQuery Unnest",
+        "description": "BigQuery",
```
Comment (william-lukusa) [inline custom-recipes/big-query-unnest/recipe.json]: TO DO : add a proper description

File: python-lib/recipe_config_loading.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    return input_dataset, output_dataset
+
+
+def parse_recipe_config(recipe_config):
+    # function to parse the UI parameters of recipe_config and put them into a params dict (or better we could do a class)
+    # can use this to validate the parameters and raise the appropriate errors to the UI
+    params = {}
+    logging.error("recipe_config:")
+    logging.error(recipe_config)
+    params["fields_to_unnest"] = recipe_config.get("fieldsToUnnest")
```
Comment (william-lukusa) [inline python-lib/recipe_config_loading.py]: Maybe we should remove empty keys here as they cause an error when running the recipe

File: resource/stylesheets/dl-image-toolbox.css [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    padding: 10px 20px;
+    border: 1px solid rgb(226, 226, 226);
+    height: 400px;
+}
+
+.several-controls .controls + .controls {
+    margin-top: 5px;
+}
+
+.mapping-editor__selector {
```
Comment (william-lukusa) [inline resource/stylesheets/dl-image-toolbox.css]: Is it ok to rename this file and remove the other unused rules ?

Comment (pbailly) [reply resource/stylesheets/dl-image-toolbox.css]: Yes !

**Review by pbailly — COMMENTED** [review]
Found issue when trying to get the following query:
```
SELECT
string_column,
dku_array_column_ AS array_column_,
array_column,
dku_array_record_column_.string_column AS scol,
dku_array_record_column_.integer_column AS icol,
record_column.datetime_column,
record_column.time_column
FROM `dataiku-dev`.`ref_data`.`all_data_types`
LEFT JOIN UNNEST(array_column) AS dku_array_column_
LEFT JOIN UNNEST(array_record_column) AS dku_array_record_column_
```

There is a mismatch on the root `record_column` and the one inside `array_record_column`. 
We should generate something like:
```
SELECT
root.string_column,
dku_array_column_ AS array_column_,
root.array_column,
dku_array_record_column_.string_column AS scol,
dku_array_record_column_.integer_column AS icol,
record_column.datetime_column,
record_column.time_column
FROM `dataiku-dev`.`ref_data`.`all_data_types` as root
LEFT JOIN UNNEST(root.array_column) AS dku_array_column_
LEFT JOIN UNNEST(root.array_record_column) AS dku_array_record_column_
```

File: python-lib/query_generator.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            select_query.append(get_select_command(field_to_unnest["path"], field_to_unnest["output"]))
+        else:
+            select_query.append(get_select_command(field_to_unnest["path"]))
+    output_query += ",\n".join(select_query) + "\n"
+
+    dataset_params = dataset.get_config()["params"]
+
+    if "catalog" in dataset_params and dataset_params["catalog"].strip():
+        if "schema" in dataset_params and dataset_params["schema"].strip():
+            output_query += "FROM `" + dataset_params["catalog"] + "`.`" + dataset_params["schema"] + "`.`" + dataset_params["table"] + "` as dku_root\n"
```
Comment (pbailly) [inline python-lib/query_generator.py:17]: Found issue when trying to get the following query:
```
SELECT
string_column,
dku_array_column_ AS array_column_,
array_column,
dku_array_record_column_.string_column AS scol,
dku_array_record_column_.integer_column AS icol,
record_column.datetime_column,
record_column.time_column
FROM `dataiku-dev`.`ref_data`.`all_data_types`
LEFT JOIN UNNEST(array_column) AS dku_array_column_
LEFT JOIN UNNEST(array_record_column) AS dku_array_record_column_
```

There is a mismatch on the root `record_column` and the one inside `array_record_column`. 
We should generate something like:
```
SELECT
root.string_column,
dku_array_column_ AS array_column_,
root.array_column,
dku_array_record_column_.string_column AS scol,
dku_array_record_column_.integer_column AS icol,
record_column.datetime_column,
record_column.time_column
FROM `dataiku-dev`.`ref_data`.`all_data_types` as root
LEFT JOIN UNNEST(root.array_column) AS dku_array_column_
LEFT JOIN UNNEST(root.array_record_column) AS dku_array_record_column_
```

Comment (pbailly) [reply python-lib/query_generator.py:17]: issue fixed

File: js/recipeController.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        $scope.callPythonDo({method}).then(function (data) {
+            updateScopeData(data);
+            $scope.finishedLoading = true;
+        }, function (data) {
+            $scope.finishedLoading = true;
+        });
+    };
+
+    this.getStylesheetUrl = function (pluginId) {
+        return `/plugins/${pluginId}/resource/stylesheets/dl-image-toolbox.css`;
```
Comment (william-lukusa) [inline js/recipeController.js:46]: should we rename this file ? I think it's name is from another plugin

**Review by pbailly — COMMENTED** [review]
Not an official LGTM, but close enough: Screenshot 2021-06-08 at 13 49 22

**Pull request merged**